### ISSUE TITLE
Difficulty aligning Rotation when importing predicted GLB & Pose to Unity3D

### ISSUE THREAD

**Issue by gaobiaoli** [facebookresearch/sam-3d-objects#91]
Hi, thanks for the great work!

I am currently trying to import the predicted GLB files and their corresponding pose information (translation, rotation, scale) into Unity3D to perfectly simulate the spatial relationship between the objects and the camera.

**The Problem:** I am struggling to find the correct coordinate transformation for the Rotation.

**Translation**: I found that simply inverting the X-axis (Right-handed to Left-handed conversion: x_unity = -x_pred) makes the position perfectly correct relative to the camera.

**Rotation**: However, no matter what coordinate transformation I attempt for the rotation (Quaternion), the orientation of the object remains incorrect.

**What I have tried**:

Using the raw output directly.

Using the provided update_layout functions.

Any advice or snippets on mapping this rotation to Unity would be greatly appreciated.

Thanks!

**Comment by gaobiaoli** [comment]
```c#
using UnityEngine;
using Newtonsoft.Json; 
using System.Collections.Generic;
using System.IO;

public class SAM3DSceneLoader : MonoBehaviour
{
    public string metadataFilePath = "unity_metadata111.json";

    public Transform objectsParent;

    [System.Serializable]
    public class ObjectData
    {
        public float[] translation;
        public float[] rotation; // [w, x, y, z]
        public float[] scale;
    }

    [ContextMenu("Load Scene Layout")]
    public void LoadLayout()
    {
        if (!File.Exists(metadataFilePath))
        {
            Debug.LogError("找不到 JSON 文件: " + metadataFilePath);
            return;
        }

        string jsonContent = File.ReadAllText(metadataFilePath);
        var data = JsonConvert.DeserializeObject<Dictionary<string, ObjectData>>(jsonContent);

        foreach (var kvp in data)
        {
            string objName = kvp.Key; 
            ObjectData objData = kvp.Value;

            Transform target = FindObjectByName(objName);
            if (target == null)
            {
                continue;
            }

            // Translation
            Vector3 rawPos = new Vector3(-objData.translation[0], objData.translation[1], objData.translation[2]);
            
            // Rotation (PyTorch [w, x, y, z])
            float w = objData.rotation[0];
            float x = objData.rotation[1];
            float y = objData.rotation[2];
            float z = objData.rotation[3];

            // Scale
            Vector3 rawScale = new Vector3(objData.scale[0], objData.scale[1], objData.scale[2]);

            
            target.position = rawPos; 

            Quaternion finalRot = new Quaternion(-x, y, z, -w);
            target.rotation = finalRot;

            target.localScale = rawScale;
        }

    }

    Transform FindObjectByName(string name)
    {
        if (objectsParent != null)
        {
            return objectsParent.Find(name);
        }
        
        GameObject obj = GameObject.Find(name);
        return obj != null ? obj.transform : null;
    }
}
```

```python
import os
import torch
from notebook.inference import Inference, ready_gaussian_for_video_rendering, load_image, load_masks, display_image, make_scene, render_video, interactive_visualizer
from sam3d_objects.pipeline.inference_utils import compose_transform,decompose_transform
import trimesh
import json
from pytorch3d.transforms import quaternion_to_matrix
from pytorch3d.transforms import Transform3d, matrix_to_quaternion

def convert_to_halo(pred_transform, device, dtype):
    on_mesh_transform = Transform3d(dtype=dtype, device=device).rotate(
        torch.tensor(
            [
                [1, 0, 0],
                [0, 0, 1],
                [0, -1, 0],
            ],
            dtype=dtype,
        )
    )
    return on_mesh_transform.compose(pred_transform) # .compose(on_pm_transform)

def update_layout(pred_t, pred_s, pred_quat, center, scale, to_halo=True):
    if center is None and not to_halo:
        return pred_t, pred_s, pred_quat
    pred_transform = compose_transform(
        pred_s, quaternion_to_matrix(pred_quat[0]), pred_t
    )
    if center is None:
        comb_transform = pred_transform
    else:
        norm_transform = compose_transform(
            scale * torch.ones_like(pred_t),
            torch.eye(3, dtype=pred_t.dtype).to(pred_t.device)[None],
            scale * -torch.tensor(center, dtype=pred_t.dtype).to(pred_t.device)[None],
        )
        comb_transform = norm_transform.compose(pred_transform)
    comb_transform = convert_to_halo(comb_transform, pred_t.device, pred_t.dtype)
    decomposed = decompose_transform(comb_transform)
    quat = matrix_to_quaternion(decomposed.rotation)
    return decomposed.translation, decomposed.scale, quat

METADATA_PATH = "/root/sam-3d-objects/glbs/nocs_0003_0354/metadata.json"
GLB_DIR = "/root/sam-3d-objects/glbs/nocs_0003_0354"
OUTPUT_DIR = "./"
os.makedirs(OUTPUT_DIR, exist_ok=True)
with open(METADATA_PATH, "r") as f:
    metadata = json.load(f)

meshes = []
final_unity_metadata = {}
for i in range(len(metadata)):
    obj = metadata[str(i)]
    glb_path = os.path.join(GLB_DIR, f"{i}.glb")
    if not os.path.exists(glb_path):
        print(f"Object {i} not found, skip")
        continue

    mesh = trimesh.load(glb_path)
    if mesh.is_empty:
        print(f"Empty mesh for {i}, skip")
        continue

    trans = torch.tensor(obj["translation"], dtype=torch.float32).reshape(1, 3)
    rotation = torch.tensor(obj["rotation"], dtype=torch.float32).reshape(1, 4)
    scale = torch.tensor(obj["scale"], dtype=torch.float32).reshape(1, 3)
    trans, scale, rotation = update_layout(trans, scale, rotation, center=None, scale=0)

    R_l2c = quaternion_to_matrix(rotation) # torch.Size([1, 3, 3])
    tfm = compose_transform(scale=scale, rotation=R_l2c, translation=trans)
    M = tfm.get_matrix()[0] # torch.Size([4, 4])
    M_trimesh = M.T.cpu().numpy() # torch.Size([4, 4])

    if isinstance(mesh, trimesh.Scene):
        for m in mesh.geometry.values():
            m.apply_transform(M_trimesh)
            meshes.append(m)
    else:
        mesh.apply_transform(M_trimesh)
        meshes.append(mesh)
    final_unity_metadata[str(i)] = {
        "translation": trans.squeeze().tolist(),
        "rotation": rotation.squeeze().tolist(), # [w, x, y, z]
        "scale": scale.squeeze().tolist()
    }

UNITY_JSON_PATH = "unity_metadata111.json"
with open(UNITY_JSON_PATH, "w") as f:
    json.dump(final_unity_metadata, f, indent=4)

print(f"Saved Unity-ready metadata to {UNITY_JSON_PATH}")
# merge all meshes
if not meshes:
    raise RuntimeError("No valid meshes to export.")

scene = trimesh.Scene()
for idx, m in enumerate(meshes):
    scene.add_geometry(m, node_name=f"obj_{idx}")

scene.export("scene111.glb") 
```

This is the code for aligning my scene with the loading code for Unity. When importing with scene111.glb, I found that it can be aligned with the image. However, when I imported each glb file into Unity and set the pose, I found that there was a rotation issue 

<img width="665" height="429" alt="Image" src="https://github.com/user-attachments/assets/826ce1e5-9359-4090-a76a-b48f000254bf" />

**Comment by alexsax** [comment]
Hi @gaobiaoli,

The Gaussians and transformations are expected to by Z-up, like in Pytorch3D conventions, while the GLB uses Y-up convention. The snippet [here](https://github.com/facebookresearch/sam-3d-objects/issues/56#issuecomment-3614031150) in #56 might be useful. I hope this helps!

**Comment by gaobiaoli** [comment]
Hi @alexsax 
Thank you for the code you provided—it gave me a completely new understanding of the coordinate systems. 
However, the issue still remains: when I import the GLB generated using your code into Unity (using the UnityGLTF plugin), I have to rotate the entire model by +90 degrees around the X-axis for it to appear in the correct orientation (I tested this using a camera with all-zero pose). I’m not sure why this adjustment is necessary.

**Comment by alexsax** [comment]
This is probably related to the coordinate convention used in Unity (e.g. see this for [unity vs blender](https://discussions.unity.com/t/handling-unity-blender-coordinate-system-differences/651032)). 

Different 3D software uses lots of different conventions -- perhaps one day the community will settle on a standard, and all our lives will be happier for it :)

**Comment by gaobiaoli** [comment]
> Hi [@gaobiaoli](https://github.com/gaobiaoli),
> 
> The Gaussians and transformations are expected to by Z-up, like in Pytorch3D conventions, while the GLB uses Y-up convention. The snippet [here](https://github.com/facebookresearch/sam-3d-objects/issues/56#issuecomment-3614031150) in [#56](https://github.com/facebookresearch/sam-3d-objects/issues/56) might be useful. I hope this helps!

What is the specific meaning of the pose in the output, and why is pytorch 3d Z-up? In its camera coordinate system, the x-axis is clearly left, y-axis is upward, and z-axis is forward

**Comment by alexsax** [comment]
You are right that the camera conventions are Z-forward, and I believe that we use that too. The pytorch3D `world` space convention is also Y-up/Z-forward.

In SAM 3D the "object space" for the Gaussians are Z-up, and the meaning of the pose is the transform from `object` space --> `camera` space. (see the [snippet](https://github.com/facebookresearch/sam-3d-objects/issues/56#issuecomment-3614031150) I mentioned before for the specifics)
